test/unit/org/apache/cassandra/metrics/CompactionMetricsTest.java

```diff
@@ -0,0 +1,229 @@
+        ActiveCompactionsTracker activeCompactions = CompactionManager.instance.active;
+        OperationType compactionType = OperationType.COMPACTION;
+        private final ColumnFamilyStore cfs;
+        private LifecycleTransaction txn;
+        private CompactionController controller;
+        private CompactionIterator ci;
+        private List<ISSTableScanner> scanners;
+
+        public TestCompactionTask(ColumnFamilyStore cfs, Set<SSTableReader> sstables)
+        {
+            this.cfs = cfs;
+            this.sstables = sstables;
+        }
+
+        public void start()
+        {
+            scanners = sstables.stream().map(SSTableReader::getScanner).collect(Collectors.toList());
+            txn = cfs.getTracker().tryModify(sstables, OperationType.COMPACTION);
+            assertNotNull(txn);
+            controller = new CompactionController(cfs, sstables, Integer.MIN_VALUE);
+            ci = new CompactionIterator(txn.opType(), scanners, controller, FBUtilities.nowInSeconds(), UUID.randomUUID());
+            activeCompactions.beginCompaction(ci);
+        }
+
+        public void start2()
+        {
+            scanners = sstables.stream().map(SSTableReader::getScanner).collect(Collectors.toList());
+            txn = cfs.getTracker().tryModify(sstables, OperationType.COMPACTION);
+            assertNotNull(txn);
+            controller = new CompactionController(cfs, sstables, Integer.MIN_VALUE);
+            ci = new CompactionIterator(txn.opType(), scanners, controller, FBUtilities.nowInSeconds(), UUID.randomUUID());
+            final Set<SSTableReader> fullyExpiredSSTables = controller.getFullyExpiredSSTables();
+            dropExpiredSSTables(fullyExpiredSSTables);
+            activeCompactions.beginCompaction(ci);
+        }
+
+        public void abort()
+        {
+            if (controller != null)
+                controller.close();
+            if (ci != null)
+                ci.close();
+            if (txn != null)
+                txn.abort();
+            if (scanners != null)
+                scanners.forEach(ISSTableScanner::close);
+            activeCompactions.finishCompaction(ci);
+        }
+
+        protected void dropExpiredSSTables(final Set<SSTableReader> fullyExpiredSSTables)
+        {
+            final Set<SSTableReader> nonExpiredSSTables = Sets.difference(txn.originals(), fullyExpiredSSTables);
+            int nonExpiredSSTablesSize = nonExpiredSSTables.size();
+            int sstablesRemoved = 0;
+
+            while (nonExpiredSSTablesSize > 1)
+            {
+                CompactionManager.instance.incrementAborted();
+                nonExpiredSSTablesSize--;
+                sstablesRemoved++;
+            }
+
+            if (sstablesRemoved > 0)
+            {
+                CompactionManager.instance.incrementCompactionsReduced();
+                CompactionManager.instance.incrementSstablesDropppedFromCompactions(sstablesRemoved);
+            }
+        }
+    }
+}
```

Comment:
Perhaps we should have another test that simply reads all related MBeans to make sure they're at the expected name and have the expected type.